rest/rest-resources/src/main/java/org/apache/brooklyn/rest/util/DefaultExceptionMapper.java

```diff
@@ -120,6 +121,32 @@ public Response toResponse(Throwable throwable1) {
         return rb.build().asResponse(Status.INTERNAL_SERVER_ERROR, MediaType.APPLICATION_JSON_TYPE);
     }
     
+    /** Logs full details at trace, unless it is the first time we've seen this in which case it is debug */
+    @Beta
+    public static void logExceptionDetailsForDebugging(Throwable t) {
+        if (LOG.isDebugEnabled()) {
+            if (firstEncounter(t)) {
+                // include debug trace for everything the first time we get one
+                LOG.debug("Full details of "+Entitlements.getEntitlementContext()+" "+t+" (logging debug on first encounter; subsequent instances will be logged trace)", t);
+            } else if (LOG.isTraceEnabled()) {
+                LOG.trace("Full details of "+Entitlements.getEntitlementContext()+" "+t, t);
+            }
+        }
+    }
+        
+    private static boolean firstEncounter(Throwable t) {
+        Set<Object> record = MutableSet.of();
+        while (true) {
```

Comment:
`while (true)` always looks alarming; in this case it may well be fine, as `t = t.getCause()` won't last forever. But why not simply write `while (t != null)` and avoid the `if (t==null) break;`?